src/app/modules/color-scheme/color-scheme-popup/color-scheme-popup.component.ts

```diff
@@ -0,0 +1,240 @@
+  positions: [0, .166, .333, .5, .666, .833, 1]
+};
+
+/**
+ * Default Scheme5 (discrete)
+ */
+const scheme5: ColorScheme = {
+  type: 'discrete',
+  name: 'purplelightblue',
+  colors: ['#6E016B', '#88419D', '#8C6BB1', '#8C96C6', '#9EBCDA', '#BFD3E6', '#EDF8FB'],
+  positions: [0, .166, .333, .5, .666, .833, 1]
+};
+
+/**
+ * Default Scheme6 (gradient)
+ */
+const scheme6: ColorScheme = {
+  type: 'gradient',
+  name: 'viridis',
+  colors: [undefined],
+  positions: [0, .5, 1]
+};
+
+/**
+ * Default Scheme7 (gradient)
+ */
+const scheme7: ColorScheme = {
+  type: 'gradient',
+  name: 'magma',
+  colors: [undefined],
+  positions: [0, .5, 1]
+};
+
+/**
+ * Default Scheme8 (gradient)
+ */
+const scheme8: ColorScheme = {
+  type: 'gradient',
+  name: 'plasma',
+  colors: [undefined],
+  positions: [0, .5, 1]
+};
+
+/**
+ * Default scheme options
+ */
+const DEFAULT_COLOR_SCHEMES =
+  [scheme1, scheme2, scheme3, scheme4, scheme5, scheme6, scheme7, scheme8];
+
+/**
+ * Component for the scheme selector popup
+ */
+@Component({
+  selector: 'ccf-color-scheme-popup',
+  templateUrl: './color-scheme-popup.component.html',
+  styleUrls: ['./color-scheme-popup.component.scss']
+})
+export class ColorSchemePopupComponent {
+
+  /**
+   * Controls visibility of the popup
+   */
+  popupVisible = false;
+
+  /**
+   * Current color scheme selected (default is scheme1)
+   */
+  colorScheme: ColorScheme = scheme1;
+
+  /**
+   * Schemes available to choose from
+   */
+  schemeOptions: ColorScheme[] = DEFAULT_COLOR_SCHEMES;
+
+  /**
+   * Current color selected (undefined for gradients)
+   */
+  color: string | undefined;
+
+  /**
+   * Brightness of selected scheme
+   */
+  brightness: number[] = [0, 100];
+
+  /**
+   * Transparency of selected scheme
+   */
+  transparency = 0;
+
+  /**
+   * Emitted when there is a color scheme change
+   */
+  @Output() colorSchemeChange = new EventEmitter<ColorScheme>();
+
+  /**
+   * Emitted when the selected color changes
+   */
+  @Output() colorChange = new EventEmitter<string | undefined>();
+
+  /**
+   * Emitted when the brightness selection changes
+   */
+  @Output() brightnessChange = new EventEmitter<number[]>();
+
+  /**
+   * Emitted when the transparency value changes
+   */
+  @Output() transparencyChange = new EventEmitter<number>();
+
+  /**
+   * Listens to document click event
+   * Closes the popup only if user clicks outside the popup
+   * @param target The element on which the event was fired
+   */
+  @HostListener('document:click', ['$event.target']) // tslint:disable-line:no-unsafe-any
+  close(target: HTMLElement): void {
+    const popupElement = document.getElementById('scheme-contents');
+    const buttonElement = document.getElementById('open-popup');
+    if (!this.popupVisible) {
+      return;
+    } else if (buttonElement && buttonElement.contains(target)) {
+      return;
+    } else if (popupElement && popupElement.contains(target)) {
+      return;
```

Comment:
Sonar complains about this. Turn into a single if statement. I.e.
```
if (cond1 ||
    (cond2 && cond3) ||
    (cond4 && cond5)) {
  return;
}
```

Reply:
I would put parenthesis around the sub conditions for clarity. Also prevents accidental errors during refactoring/updating. I.e.
`... || (buttonElement && buttonElement.contains(target)) || ...`